File: cpp/src/asset_context.cpp
#include "asset_context.hpp"
#include <algorithm>
#include <iterator>

namespace longbridge {
namespace asset {

AssetContext::AssetContext()
  : ctx_(nullptr)
{
}

AssetContext::AssetContext(const lb_asset_context_t* ctx)
{
  ctx_ = ctx;
  if (ctx_) {
    lb_asset_context_retain(ctx_);
  }
}

AssetContext::AssetContext(const AssetContext& ctx)
{
  ctx_ = ctx.ctx_;
  if (ctx_) {
    lb_asset_context_retain(ctx_);
  }
}

AssetContext::AssetContext(AssetContext&& ctx)
{
  ctx_ = ctx.ctx_;
  ctx.ctx_ = nullptr;
}

AssetContext::~AssetContext()
{
  if (ctx_) {
    lb_asset_context_release(ctx_);
  }
}

AssetContext&
AssetContext::operator=(const AssetContext& ctx)
{
  ctx_ = ctx.ctx_;
  if (ctx_) {
    lb_asset_context_retain(ctx_);
  }
  return *this;
}

AssetContext
AssetContext::create(const Config& config)
{
  auto* ctx_ptr = lb_asset_context_new(config);
  AssetContext ctx(ctx_ptr);
  if (ctx_ptr) {
    lb_asset_context_release(ctx_ptr);
  }
  return ctx;
}

void
AssetContext::statements(
  int32_t statement_type,
  int32_t start_date,
  int32_t limit,
  AsyncCallback<AssetContext, std::vector<StatementItem>> callback) const
{
  lb_asset_context_statements(
    ctx_,
    statement_type,
    start_date,
    limit,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<AssetContext,
                                     std::vector<StatementItem>>(
          res->userdata);
      AssetContext ctx((const lb_asset_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        auto rows = (const lb_statement_item_t*)res->data;
        std::vector<StatementItem> rows2;
        std::transform(rows,
                       rows + res->length,
                       std::back_inserter(rows2),
                       [](const auto& row) {
                         StatementItem item;
                         item.dt = row.dt;
                         item.file_key = row.file_key;
                         return item;
                       });

        (*callback_ptr)(
          AsyncResult<AssetContext, std::vector<StatementItem>>(
            ctx, std::move(status), &rows2));
      } else {
        (*callback_ptr)(
          AsyncResult<AssetContext, std::vector<StatementItem>>(
            ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<AssetContext, std::vector<StatementItem>>(callback));
}

void
AssetContext::statement_download_url(
  const std::string& file_key,
  AsyncCallback<AssetContext, StatementDownloadUrlResponse> callback) const
{
  lb_asset_context_download_url(
    ctx_,
    file_key.c_str(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<AssetContext,
                                     StatementDownloadUrlResponse>(
          res->userdata);
      AssetContext ctx((const lb_asset_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        auto resp = (const lb_statement_download_url_response_t*)res->data;
        StatementDownloadUrlResponse result;
        result.url = resp->url;

        (*callback_ptr)(
          AsyncResult<AssetContext, StatementDownloadUrlResponse>(
            ctx, std::move(status), &result));
      } else {
        (*callback_ptr)(
          AsyncResult<AssetContext, StatementDownloadUrlResponse>(
            ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<AssetContext, StatementDownloadUrlResponse>(
      callback));
}

} // namespace asset
} // namespace longbridge

File: java/src/asset_context.rs
use std::sync::Arc;

use jni::{
    JNIEnv,
    objects::{JClass, JObject},
};
use longbridge::{
    Config,
    asset::{AssetContext, GetStatementListOptions, GetStatementOptions, StatementType},
};

use crate::{
    async_util,
    error::jni_result,
    types::{FromJValue, JavaInteger, get_field},
};

struct ContextObj {
    ctx: AssetContext,
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_newAssetContext(
    mut env: JNIEnv,
    _class: JClass,
    config: i64,
) -> i64 {
    jni_result(&mut env, 0i64, |_env| {
        let config = Arc::new((*(config as *const Config)).clone());
        let ctx = AssetContext::new(config);
        Ok(Box::into_raw(Box::new(ContextObj { ctx })) as i64)
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_freeAssetContext(
    _env: JNIEnv,
    _class: JClass,
    ctx: i64,
) {
    let _ = Box::from_raw(ctx as *mut ContextObj);
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_assetContextStatements(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let statement_type: Option<JavaInteger> = get_field(env, &opts, "statementType")?;
        let start_date: Option<JavaInteger> = get_field(env, &opts, "startDate")?;
        let limit: Option<JavaInteger> = get_field(env, &opts, "limit")?;

        let st = match statement_type.map(i32::from).unwrap_or(1) {
            2 => StatementType::Monthly,
            _ => StatementType::Daily,
        };
        let mut options = GetStatementListOptions::new(st);
        if let Some(sd) = start_date {
            options = options.page(i32::from(sd));
        }
        if let Some(l) = limit {
            options = options.page_size(i32::from(l));
        }

        async_util::execute(env, callback, async move {
            let resp = context.ctx.statements(options).await?;
            Ok(serde_json::to_string(&resp).unwrap_or_default())
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_assetContextDownloadUrl(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    file_key: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let file_key: String = FromJValue::from_jvalue(env, file_key.into())?;
        let options = GetStatementOptions::new(file_key);

        async_util::execute(env, callback, async move {
            let resp = context.ctx.statement_download_url(options).await?;
            Ok(resp.url)
        })?;
        Ok(())
    })
}
